src/rust/lib/frp/src/io/keyboard.rs

```diff
@@ -179,138 +154,119 @@ impl Default for Keyboard {
+        Actions{action_map,network}
     }
 
-    fn perform_action_lambda(action_map:Rc<RefCell<ActionMap>>) -> impl Fn(&KeyMask) {
+    fn perform_action_fn(action_map:Rc<RefCell<ActionMap>>) -> impl Fn(&KeyMask) {
         move |key_mask| {
-            let entry_opt = with(action_map.borrow_mut(), |mut map| map.remove_entry(key_mask));
-            if let Some((map_mask, mut action)) = entry_opt {
-                action(key_mask);
-                if let Entry::Vacant(entry) =  action_map.borrow_mut().entry(map_mask) {
-                    entry.insert(action);
+            // The action map ref is cloned in order to execute callbacks when not being borrowed.
+            let opt_callbacks = action_map.borrow().get(key_mask).map(|t| t.clone_ref());
+            if let Some(callbacks) = opt_callbacks {
+                callbacks.run_all();
+                if callbacks.is_empty() {
+                    action_map.borrow_mut().remove(key_mask);
                 }
             }
         }
     }
 
     /// Set action binding for given key mask.
-    pub fn set_action<F:FnMut(&KeyMask) + 'static>(&mut self, key_mask:KeyMask, action:F) {
-        self.action_map.borrow_mut().insert(key_mask,Box::new(action));
+    pub fn add_action_for_key_mask<F:Action>(&self, key_mask:KeyMask, action:F) -> callback::Handle {
+        self.action_map.borrow_mut().entry(key_mask).or_insert_with(default).add(action)
     }
 
-    /// Remove action binding for given key mask.
-    pub fn unset_action(&mut self, key_mask:&KeyMask) {
-        self.action_map.borrow_mut().remove(key_mask);
+    /// Set action binding for given set of keys.
+    pub fn add_action<F:Action>(&self, keys:&[Key], action:F) -> callback::Handle {
+        self.add_action_for_key_mask(keys.into(),action)
     }
 }
 
-impl Debug for KeyboardActions {
+impl Debug for Actions {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "<KeyboardActions>")
+        write!(f, "<CallbackRegistry>")
     }
 }
 
 
-#[cfg(test)]
-mod test {
-    use super::*;
 
-    #[test]
-    fn key_mask() {
-        let keyboard                  = Keyboard::default();
-        let expected_key_mask:KeyMask = default();
-        assert_eq!(expected_key_mask, keyboard.key_mask.value());
-        let key1 = Key::Character("x".to_string());
-        let key2 = Key::Control;
-
-        keyboard.on_pressed.emit(key1.clone());
-        let expected_key_mask:KeyMask = std::iter::once(&key1).collect();
-        assert_eq!(expected_key_mask, keyboard.key_mask.value());
-
-        keyboard.on_pressed.emit(key2.clone());
-        let expected_key_mask:KeyMask = [&key1,&key2].iter().cloned().collect();
-        assert_eq!(expected_key_mask, keyboard.key_mask.value());
-
-        keyboard.on_released.emit(key1.clone());
-        let expected_key_mask:KeyMask = std::iter::once(&key2).collect();
-        assert_eq!(expected_key_mask, keyboard.key_mask.value());
-    }
```

Comment:
I know that `Actions` are deprecated, but this test have none to do with those: it tested the `Keyboard` structure, thus we should not just remove it without any new test.